TIP-1028: Address-Level Receive Policies (#3173)
tempoxyz/tempo: tips/tip-1028.md

event TokenSetBlacklistUpdated(uint64 indexed tokenSetId, address indexed updater, address indexed token, bool restricted);

error TokenSetNotFound();
error InvalidTokenSetType();
```

## 3. Address-Level Receive Controls

Any address MAY configure four fields:

1. `receivePolicyId` — which originators may credit the address
2. `tokenSetId` — which TIP-20 token addresses may credit the address
3. `recoveryMode` — who may later claim blocked funds
4. `claimerWhitelistId` — optional TIP-403 whitelist for third-party claimers

If an address has no configured receive controls, address-level authorization defaults to allow.

TIP-1022 virtual addresses are forwarding aliases, not canonical TIP-20 holders. `setAddressReceivePolicy()` MUST reject TIP-1022 virtual addresses and require configuration on the resolved master address instead.

### 3.1 Constraints

`receivePolicyId` MUST reference:

- a simple `WHITELIST` policy,
- a simple `BLACKLIST` policy, or
- built-in policy `0` or `1`

It MUST NOT reference a `COMPOUND` policy.

`claimerWhitelistId` MUST reference:

- a simple `WHITELIST` policy, or
- built-in policy `0` or `1`

It MUST NOT reference a `BLACKLIST` or `COMPOUND` policy.

> ### 3.2 Recovery Modes

Reviewer: Can we just have the user specify a recovery contract, and that recovery contract is what can call the escrow address to claim the tokens? Then we don't need a bunch of manually specified recovery modes

Reviewer 2: This is nice, this also allows for recovery contract upgrades in the future without having.

- `(false, TOKEN_SET_AND_RECEIVE_POLICY)` if both checks fail
- `(false, TOKEN_SET)` if only the token-set check fails
- `(false, RECEIVE_POLICY)` if only the receive-policy check fails

An address that wants to functionally disable filtering SHOULD set `receivePolicyId = 1` and `tokenSetId = 1`. The slot remains allocated.

### 3.6 Events and Errors

```solidity
event AddressReceivePolicyUpdated(
    address indexed account,
    uint64 receivePolicyId,
    uint64 tokenSetId,
    address recoveryContract
);

error InvalidReceivePolicyType();
error InvalidRecoveryContract();
```

## 4. Escrow Precompile

Blocked inbounds are recorded in a dedicated escrow precompile. The raw TIP-20 balance is held at `ESCROW_ADDRESS` inside each TIP-20 token; the precompile stores only one keyed amount for each open blocked receipt.

```solidity
ESCROW_ADDRESS = 0xE5C0000000000000000000000000000000000000
```

### 4.1 Storage

Each blocked inbound creates exactly one fine-grained receipt bucket.

```solidity
uint64 public blockedReceiptNonce = 1;
mapping(bytes32 => uint256) internal blockedReceiptAmount;
```

The persistent escrow key for a blocked inbound is:

```text

> receiptKey = keccak256(

Reviewer 3: add domain separator / version number

---
id: TIP-1028
title: Address-Level Receive Policies
description: Extends TIP-403 with token sets, address-level receive controls, and receipt-based escrow for blocked TIP-20 inbounds.
authors: Mallesh Pai, 0xrusowsky
status: Draft
related: TIP-403, TIP-1015, TIP-20, TIP-1016, TIP-1022
protocolVersion: TBD
---

# TIP-1028: Address-Level Receive Policies

## Abstract

This TIP extends TIP-403 with token sets for TIP-20 token addresses, address-level receive controls, and non-reverting escrow for blocked TIP-20 inbounds. A blocked inbound credits `ESCROW_ADDRESS` and creates one fine-grained receipt bucket.

> The escrow precompile stores only one keyed amount for each open blocked receipt. The rest of the receipt identity — receipt version, requested recipient, block reason, transfer-vs-mint kind, memo, timestamp, and nonce — is authenticated by the receipt witness and emitted in the blocked event. Claims may unwind to the receiver or reroute elsewhere; claim-to-receiver is an unwind, while reroutes are new spends. This TIP applies only to TIP-20 precompile flows; ordinary ERC-20 contracts remain out of scope.

Reviewer 2: nit: the abstract jumps into details like receipt version, transfer-vs-mint kind, etc. before the reader actually is introduced to this context. I think the abstract should be a bit higher level. 

```suggestion
TIP-1028 extends TIP-403 with address-level receive policies and token sets, allowing a receiver to control which originators and which TIP-20 tokens may credit it. When a receive policy blocks an inbound transfer or mint, the operation still succeeds at the protocol level, but the funds are credited to `ESCROW_ADDRESS` instead of the receiver and a corresponding escrow receipt is recorded.

The receiver or a designated recovery contract may later claim these escrowed funds. Claims back to the receiver act as an unwind of the original inbound, while claims to other addresses are treated as new transfers.

Token-level authorization remains unchanged and continues to revert on failure. This behavior applies only to TIP-20 precompile flows; ordinary contracts and other precompiles are unaffected.
```

## Motivation

TIP-403 lets token issuers decide who may use a token, but some receivers also need inbound controls. A revert-based receiver policy creates a liveness problem: once a relationship exists, the receiver can later change policy and cause future transfers or mints to revert.

> This TIP keeps issuer-side semantics unchanged and changes only receiver-side failure handling. Token-level failures still revert; receiver-side failures are escrowed instead. The escrow representation must preserve more than an aggregate amount because TIP-1022 virtual addresses carry attribution in the literal `to` address, memo-bearing transfers carry memo data, recovery contracts may want time-based or memo-based rules, and offchain recovery flows need the original sender identity.

Reviewer 2: ```suggestion
TIP-403 allows token issuers to control who may use a token, but it does not give receivers control over which inbounds they accept. Some applications require receivers to restrict incoming transfers or mints based on the originator or token.

A revert-based receiver policy introduces a liveness problem: once a sender relationship exists, a receiver can later change its policy and cause future transfers or mints to revert, breaking integrations and making delivery unreliable.

This TIP introduces receiver-side controls while preserving liveness. Instead of reverting when a receiver blocks an inbound, the transfer or mint succeeds at the protocol level and the funds are escrowed. This ensures that senders and integrations are not disrupted, while still giving receivers control over acceptance.

Blocked inbounds must remain recoverable and attributable, allowing the escrow model to preserve per-transfer context such as originator, requested recipient, and metadata, enabling flexible recovery and offchain handling.
```

- `systemTransferFrom`
- `mint`
- `mintWithMemo`
- protocol withdrawals that execute as TIP-20 transfers from a concrete source balance

For all such paths, TIP-1028 adds a receiver-side authorization layer:

- the receiver's token set is checked against the TIP-20 token address; and
- the receiver's receive policy is checked against the inbound **originator**:
  - `from` for transfer-like paths;
  - the mint caller for `mint` and `mintWithMemo`.

If a token-level TIP-403 or TIP-1015 policy rejects the operation, the operation MUST revert exactly as it does today. If the receiver's address-level controls reject the inbound, the operation MUST be escrowed instead.

If `to` is a TIP-1022 virtual address, TIP-1022 recipient resolution MUST occur before TIP-1028 receiver-side authorization. In that case:

- TIP-1028 applies to the resolved master address, not the literal virtual address;
- if virtual-address resolution fails, the operation MUST revert rather than escrow;
- the blocked receipt's `receiver` MUST be the resolved master address;
- the blocked receipt's `requestedRecipient` MUST be the literal `to` address so offchain systems can recover the TIP-1022 `userTag`; and

> - if the inbound is authorized, the success path MUST then follow TIP-1022 forwarding semantics.

Reviewer 2: This is talking about "blocked receipt" before its introduced/defined. Virtual addresses should be reintroduced when talking about the claim path in `Claim Authorization` or `Release Semantics`. I would remove it from here and simplify. 

```suggestion
If `to` is a TIP-1022 virtual address, TIP-1022 recipient resolution MUST occur before TIP-1028 receiver-side authorization. If virtual-address resolution fails, the operation MUST revert.
```

> TIP-1028 does **not** alter `approve`, `permit`, `burn`, fee refunds via `transfer_fee_post_tx`, non-TIP-20 tokens deployed as ordinary contracts, or future recipient-bearing system-credit paths that do not identify a concrete originator.

Reviewer 2: > non-TIP-20 tokens deployed as ordinary contracts

I dont think its necessary to mention this again. We should mention that "TIP-1028 only applies to TIP-20 precompile flows; ordinary contracts and other precompiles are unaffected" once in the overview, rather than restate this.

> DEX internal balances are not TIP-20 wallet balances and are not subject to address-level receive checks until withdrawn back onto the TIP-20 ledger.

Reviewer 2: Similar to the comment above, I dont think we need to mention this as this is implied when stating that only TIP-20 precompiles are affected.

The reward subsystem is not receiver-side escrowable, but it is not exempt from recipient consent. `distributeReward` remains token-internal accounting; `setRewardRecipient` MUST reject a nonzero recipient whose address-level receive controls would block the holder as originator; and `claimRewards` MUST revalidate the actual payout recipient's address-level receive controls against the token contract as originator and MUST revert on failure rather than escrow.

> For reward accounting, `ESCROW_ADDRESS` is a reward-exempt always-opted-out synthetic sink/source: blocked transfers, blocked mints, and claim releases MUST preserve the same opted-in-supply effects as a movement into or out of an always-opted-out address, and implementations MUST NOT create, update, or consult per-user reward state for `ESCROW_ADDRESS`.

Reviewer 2: This feels a bit over-specified, once the scope is limited to recipient-bearing TIP-20 ledger flows, internal accounting flows like rewards or DEX internal balances should already be implied out of scope. We should avoid enumerating every internal balance system unless there’s a non-obvious exception.

If we want to keep something to explictly callout rewards since they are part of TIP20s, I would suggest that we update to:

```suggestion
Reward flows are not subject to escrow. Reward payouts MUST still respect address-level receive controls and MUST revert if the recipient would reject the inbound.
```

High-level flow:

1. If `to` is a TIP-1022 virtual address, resolve it first to `effectiveReceiver`; otherwise `effectiveReceiver = to`.

> 2. Run the existing TIP-20 sender-side or issuer-side checks and the token's TIP-403 and TIP-1015 checks using `effectiveReceiver` wherever the current path is recipient-sensitive.

Reviewer 2: nit:
```suggestion
2. Run existing TIP-20 sender/issuer checks, including TIP-403 and TIP-1015, using `effectiveReceiver` where applicable.
```

3. Run the receiver's token-set and receive-policy checks against `effectiveReceiver`.
4. If the inbound is authorized, credit the receiver normally.
5. If the inbound is blocked by the receiver, credit `ESCROW_ADDRESS`, create one blocked receipt, and emit a blocked-inbound event naming both `receiver` and `requestedRecipient`.
6. Later, the receiver or the receipt's recovery contract may claim the funds.

> 

Reviewer 2: It would be nice to have a simple sequence daigram demonstrating the high level flow here.

```suggestion

Below is a simplified sequence diagram showing the TIP-20 inbound flow, including escrow of blocked inbounds and how funds are claimed.

sequenceDiagram
    actor Sender
    participant TIP20 as TIP-20
    participant Escrow
    actor Receiver

    Sender->>TIP20: transfer / mint to Receiver
    TIP20->>TIP20: resolve receiver and evaluate inbound rules
    TIP20->>Escrow: record blocked inbound receipt
    TIP20->>TIP20: credit ESCROW_ADDRESS
    TIP20-->>Sender: success

    Receiver->>Escrow: claim blocked receipt
    Escrow->>TIP20: release from ESCROW_ADDRESS
    TIP20-->>Receiver: credit claimed funds

```

## 2. Token Sets

Token sets are a dedicated TIP-403 primitive for TIP-20 token addresses. They answer a different question from address policies:

- address policy: "is this originator authorized?"
- token set: "is this token authorized?"

> Token sets use a separate ID space from policy IDs. They are not aliases for ordinary TIP-403 policy lists and do not reuse the compound-policy surface, but they mirror the ordinary TIP-403 list ergonomics, including create-with-members and batched membership updates.

Reviewer 2: It feels like token sets are introduced too early without context. The core mental model of this TIP is the inbound flow (credit vs escrow), it would be clearer to introduce escrow and the overall flow first, then define token sets and receive policies as part of the receiver-side authorization step.

### 2.1 Storage and Constraints

```solidity
uint64 public tokenSetIdCounter = 2; // 0 = reject all, 1 = allow all

struct TokenSetData {
    PolicyType setType; // WHITELIST or BLACKLIST
    address admin;
}

mapping(uint64 => TokenSetData) internal _tokenSetData;
mapping(uint64 => mapping(address => bool)) internal tokenSetMembers;
```

Built-in meanings: `0` = always reject; `1` = always allow.

> Token sets MUST satisfy: `setType` is `WHITELIST` or `BLACKLIST`; `COMPOUND` token sets are forbidden; token-set type is immutable after creation; and membership is mutable by the token-set admin.

Reviewer 2: Similar to the comment above, storage and constraints are introduced without actually introducing what contract this relates to. Would should introduce the escrow precompile first to make it easier to build a mental model of the subsequent moving parts.

### 2.2 Interface

```solidity
interface ITIP403TokenSets {
    function createTokenSet(address admin, PolicyType setType)
        external
        returns (uint64 newTokenSetId);

    function createTokenSetWithTokens(
        address admin,
        PolicyType setType,
        address[] calldata tokens

>     ) external returns (uint64 newTokenSetId);

Reviewer 2: It’s not clear what contract implements the `ITIP403TokenSets` interface. Im under the impression it should be the TIP403 precompile but the spec does not explicitly state this. Would be good to make this explicit to avoid confusion.

### 2.5 Events and Errors

```solidity
event TokenSetCreated(uint64 indexed tokenSetId, address indexed creator, PolicyType setType);
event TokenSetAdminUpdated(uint64 indexed tokenSetId, address indexed updater, address indexed admin);
event TokenSetWhitelistUpdated(uint64 indexed tokenSetId, address indexed updater, address indexed token, bool allowed);
event TokenSetBlacklistUpdated(uint64 indexed tokenSetId, address indexed updater, address indexed token, bool restricted);

error TokenSetNotFound();
error InvalidTokenSetType();
error TokenSetBatchLengthMismatch();
```

## 3. Address-Level Receive Controls

Any address MAY configure three fields: `receivePolicyId` for which originators may credit the address, `tokenSetId` for which TIP-20 token addresses may credit the address, and `recoveryContract` for an optional claimer authorized to recover blocked receipts one receipt at a time; `address(0)` means the receiver claims directly. If an address has no configured receive controls, address-level authorization defaults to allow.

> TIP-1022 virtual addresses are forwarding aliases, not canonical TIP-20 holders. `setAddressReceivePolicy()` MUST reject TIP-1022 virtual addresses and require configuration on the resolved master address instead.

Reviewer 2: “Any address MAY configure…” is unclear. Where is this configuration stored and which contract exposes `setAddressReceivePolicy()`? It would help to explicitly state which component owns address-level receive controls. At the moment, the reader has to infer it which is confusing.

| `144..254` | 111 | reserved, MUST be zero |
| `255` | 1 | `hasAddressPolicy` |

When `hasAddressPolicy == 0`, the address is always authorized at the address level. The cached type fields are valid because policy type and token-set type are immutable after creation.

`recoveryContract` is stored separately because a 160-bit address does not fit in the packed config slot.

### 3.4 Interface

```solidity
interface IAddressReceivePolicies {
    function setAddressReceivePolicy(
        uint64 receivePolicyId,
        uint64 tokenSetId,
        address recoveryContract
    ) external;

    function addressReceivePolicy(address account)
        external
        view
        returns (
            bool hasAddressPolicy,
            uint64 receivePolicyId,
            PolicyType receivePolicyType,
            uint64 tokenSetId,
            PolicyType tokenSetType,
            address recoveryContract
        );

    function classifyAddressInbound(address token, address originator, address to)
        external
        view
        returns (
            bool authorized,
            address recoveryContract,
            BlockedReason blockedReason

>         );

Reviewer 4: i think this fn is overscoped, since the `recoveryContract` info adds an extra SLOAD and that info is already available via `addressReceivePolicy`.

if we think it's still useful we can keep it, but IMO the API only needs a minimal version that allows contracts to pre-verify whether a transfer will succeed (without the extra SLOAD for the recovery contract):

```solidity
verifyAddressInbound(address token, address originator, address to)
        external
        view
        returns (bool authorized, BlockedReason blockedReason);
```

Reviewer 3: done

}
```

Changing `recoveryContract` affects only future blocked buckets. Existing buckets remain governed by the `recoveryContract` recorded when they were blocked.

> ### 3.5 Authorization Logic

Reviewer 4: if we adopt the proposed `verifyAddressInbound`, this section should be updated to use that fn. Additionally, it should explicitly say that impl should only read `addressRecoveryContract[to]` if the inbound is actually blocked

Reviewer 3: done

- if both checks pass, return:
  - `authorized = true`
  - the current `recoveryContract`
  - `blockedReason = NONE`
- if both checks fail, return:
  - `authorized = false`
  - the current `recoveryContract`
  - `blockedReason = TOKEN_SET_AND_RECEIVE_POLICY`
- if only the token-set check fails, return:
  - `authorized = false`
  - the current `recoveryContract`
  - `blockedReason = TOKEN_SET`
- if only the receive-policy check fails, return:
  - `authorized = false`
  - the current `recoveryContract`
  - `blockedReason = RECEIVE_POLICY`

An address that wants to functionally disable filtering SHOULD set `receivePolicyId = 1` and `tokenSetId = 1`. The slot remains allocated.

### 3.6 Events and Errors

```solidity
event AddressReceivePolicyUpdated(
    address indexed account,
    uint64 receivePolicyId,
    uint64 tokenSetId,
    address recoveryContract
);

error InvalidReceivePolicyType();
error InvalidRecoveryContract();
```

## 4. Escrow Precompile

Blocked inbounds are recorded in a dedicated escrow precompile. The raw TIP-20 balance is held at `ESCROW_ADDRESS` inside each TIP-20 token; the precompile stores receipt metadata and claimability.

```solidity

> ESCROW_ADDRESS = 0xFDC1000000000000000000000000000000000000

Reviewer 3: 0xE5c00000
